File: src/soulmap/devtools/web/content.py
"""Load canonical repository Markdown into the typed public page model.

This is the only place that reads `skills/` for the website, and it enforces
the allowlist while it reads. A document that is not public never becomes a
page model at all, and an internal section never enters one, so no downstream
template can leak either.

Markdown rendering uses `markdown-it-py`, already a main dependency of this
package, and the shared front-matter and anchor helpers in
`soulmap.devtools.support.markdown`, so the website never grows a second
Markdown implementation that could drift from the repository's own.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path

from markdown_it import MarkdownIt

from soulmap.devtools.support.markdown import (
    parse_yaml_front_matter,
    slugify_github_anchor,
)
from soulmap.devtools.web.allowlist import is_internal_section, is_public_skill

_ATX_HEADING_RE = re.compile(r"^(#{1,6})\s+(.*)$")
_FENCE_RE = re.compile(r"^\s*(`{3,}|~{3,})")

# A bold label standing in for a heading: **Detection signals:** or
# **Activation Signals**. Matched case-insensitively against the same names as
# the heading form, so both spellings are covered by one list.
_MD_LINK_RE = re.compile(r"\[([^\]]+)\]\(([^)\s]+)\)")

# Path segments belonging to material the website does not publish. Matching on
# segments rather than on a full prefix matters: canonical files link to each
# other relatively, so the same private target appears as `skills/meta/x.md`,
# `../meta/x.md`, and `./meta/x.md` in different files.
_PRIVATE_PATH_SEGMENTS: frozenset[str] = frozenset(
    {"meta", "safety", "spiritual", "soulmate", "templates", ".claude", "src"}
)


def _is_private_target(reference: str) -> bool:
    """Report whether a link target or label points at non-public material.

    Args:
        reference: A link target or its visible label.

    Returns:
        True when any path segment names a private area, or the reference
        points at a category ``SKILL.md``, which the allowlist never publishes.
    """
    cleaned = reference.split("#", 1)[0].strip()
    segments = [part for part in cleaned.split("/") if part not in {"", ".", ".."}]
    if any(segment in _PRIVATE_PATH_SEGMENTS for segment in segments[:-1]):
        return True
    return cleaned.endswith("SKILL.md")


_LABELLED_SIGNAL_RE = re.compile(
    r"^\*\*\s*(detection signals?|activation signals?|paired template)\s*:?\s*\*\*:?\s*$",
    re.IGNORECASE,
)

# The priority table in SOULMAP.md, one framework per row:
# | Medium | Inner Parts | Inner conflict is present without clear insight |
_PRIORITY_ROW_RE = re.compile(
    r"^\|\s*(Highest|Very high|High|Medium|Lower|Default)\s*\|\s*([^|]+?)\s*\|"
)

TIER_ORDER: tuple[str, ...] = (
    "Highest",
    "Very high",
    "High",
    "Medium",
    "Lower",
    "Default",
)


@dataclass(frozen=True, slots=True)
class PublicSection:
    """One rendered section of a public page.

    Attributes:
        heading: Section heading text, without leading hashes.
        level: Heading level, 2 for ``##``.
        anchor: GitHub-compatible slug, for in-page links.
        html: Rendered body HTML, excluding the heading itself.
    """

    heading: str
    level: int
    anchor: str
    html: str


@dataclass(frozen=True, slots=True)
class PublicPage:
    """One public page derived from a canonical repository document.

    Attributes:
        slug: URL segment, derived from the filename stem.
        name: Human-readable name, from front matter or the first heading.
        description: One-line summary from front matter.
        category: Parent directory under `skills/`, or ``"doctrine"``.
        tier: Priority tier from the SOULMAP.md table, None when the document
            is not a routed framework.
        sections: Public sections in document order, internal ones removed.
        source_path: Repository-relative path, for the "read the source" link.
    """

    slug: str
    name: str
    description: str
    category: str
    tier: str | None
    sections: tuple[PublicSection, ...]
    source_path: str


def _delink_private_targets(text: str) -> str:
    """Turn links to non-public files into plain text.

    An allowlisted document may still reference an internal one. Rendering that
    reference as a link would emit a dead link and advertise the internal
    file's path. The reference is kept as plain text so the sentence still
    reads, but it stops being a pointer, and the path itself is dropped.

    Args:
        text: Markdown body text.

    Returns:
        The text with private-target links flattened to their label.
    """

    def replace(match: re.Match[str]) -> str:
        label, target = match.group(1), match.group(2)
        if _is_private_target(target) or _is_private_target(label):
            # A bare path used as its own label would leave the internal path
            # visible as text even after the link is removed.
            if _is_private_target(label):
                return "the internal knowledge base"
            return label
        return match.group(0)

    return _MD_LINK_RE.sub(replace, text)


def _markdown_renderer() -> MarkdownIt:
    """Build the Markdown renderer used for every public page.

    Returns:
        A CommonMark renderer with tables enabled and raw HTML disabled.
        Disabling raw HTML matters: canonical files are trusted, but rendering
        them with HTML passthrough would make any future authoring mistake a
        markup injection into the public site.
    """
    md = MarkdownIt("commonmark", {"html": False, "linkify": False})
    md.enable("table")
    return md


def _split_front_matter(text: str) -> tuple[dict[str, str], list[str]]:
    """Separate YAML front matter from the document body.

    Args:
        text: The full file contents.

    Returns:
        The parsed front-matter mapping (empty when absent) and the body lines
        with the front-matter block removed.
    """
    lines = text.splitlines()
    front = parse_yaml_front_matter(lines) or {}
    if not front or not lines or lines[0].strip() != "---":
        return front, lines

    for index in range(1, len(lines)):
        if lines[index].strip() == "---":
            return front, lines[index + 1 :]
    return front, lines


def _iter_blocks(body: list[str]) -> list[tuple[int, str, list[str]]]:
    """Split a document body into blocks at every heading level.

    Splitting at all six levels, rather than only at ``#`` and ``##``, is a
    safety requirement rather than a formatting preference: a `### Detection
    signals` subsection must be droppable on its own, and a splitter that only
    saw level-2 headings would leave it embedded in its parent section's body.

    Args:
        body: Document lines with front matter already removed.

    Returns:
        Tuples of heading level, heading text, and body lines. The preamble
        before the first heading is returned at level 0 with an empty heading.
        Fenced code blocks are tracked so a ``#`` comment inside one is never
        mistaken for a heading.
    """
    blocks: list[tuple[int, str, list[str]]] = []
    heading = ""
    level = 0
    buffer: list[str] = []
    fence: str | None = None

    for raw in body:
        fence_match = _FENCE_RE.match(raw)
        if fence_match:
            marker = fence_match.group(1)[0]
            if fence is None:
                fence = marker
            elif fence == marker:
                fence = None
            buffer.append(raw)
            continue

        heading_match = None if fence else _ATX_HEADING_RE.match(raw)
        if heading_match:
            blocks.append((level, heading, buffer))
            level = len(heading_match.group(1))
            heading = heading_match.group(2).strip()
            buffer = []
            continue

        buffer.append(raw)

    blocks.append((level, heading, buffer))
    return blocks


def _drop_internal_blocks(
    blocks: list[tuple[int, str, list[str]]],
) -> list[tuple[int, str, list[str]]]:
    """Remove internal sections and everything nested beneath them.

    Dropping a heading must also drop its subsections, otherwise a
    `## Detection signals` section with `### Acute` beneath it would lose only
    its own preamble and publish the phrase lists underneath.

    Args:
        blocks: Blocks from ``_iter_blocks``.

    Returns:
        The blocks that may become public sections.
    """
    kept: list[tuple[int, str, list[str]]] = []
    skip_above: int | None = None

    for level, heading, lines in blocks:
        if skip_above is not None:
            if level > skip_above:
                continue
            skip_above = None

        if heading and is_internal_section(heading):
            skip_above = level
            continue

        kept.append((level, heading, lines))

    return kept


def _scrub_labelled_signals(lines: list[str]) -> list[str]:
    """Remove bold-labelled signal blocks that are not Markdown headings.

    Several framework files mark phrase lists with a bold label rather than a
    heading, for example ``**Detection signals:**`` followed by a bullet list.
    `pattern-mapper.md` alone carries six. A heading-only filter leaves every
    one of them public, so the label form is stripped here with the bullet list
    that follows it.

    Args:
        lines: A block's body lines.

    Returns:
        The lines with labelled signal blocks removed.
    """
    kept: list[str] = []
    dropping = False

    for raw in lines:
        stripped = raw.strip()

        if _LABELLED_SIGNAL_RE.match(stripped):
            dropping = True
            continue

        if dropping:
            # Consume the label's bullet list and any blank lines inside it.
            # The first line that is neither blank nor a list item ends it.
            if not stripped or stripped.startswith(("-", "*", "+")):
                continue
            dropping = False

        kept.append(raw)

    return kept


def load_priority_tiers(repo_root: Path) -> dict[str, str]:
    """Read the framework priority tiers from the SOULMAP.md doctrine table.

    The website groups frameworks the way SoulMap actually decides between
    them, so the tiers are parsed from doctrine rather than hardcoded. When the
    table changes, the site regroups itself on the next build.

    Args:
        repo_root: Repository root.

    Returns:
        A mapping of lowercased framework display name to its priority tier.

    Raises:
        OSError: If SOULMAP.md cannot be read. It is canonical doctrine, so a
            missing file is a repository error rather than a condition to
            absorb.
    """
    text = (repo_root / "SOULMAP.md").read_text(encoding="utf-8")
    tiers: dict[str, str] = {}
    for line in text.splitlines():
        match = _PRIORITY_ROW_RE.match(line)
        if match:
            tiers[match.group(2).strip().lower()] = match.group(1).strip()
    return tiers


# Words that stay lowercase inside a title unless they lead it, so a slug
# fallback does not produce "Dark Night Of Soul".
_TITLE_MINOR_WORDS: frozenset[str] = frozenset(
    {"a", "an", "and", "as", "at", "by", "for", "in", "of", "on", "or", "the", "to"}
)


def _display_name(slug: str) -> str:
    """Turn a filename stem into a readable title.

    This is a last-resort fallback. Every current knowledge file carries an H1,
    so the document's own title is used instead, and that title is canonical
    where this rule would only be a guess.

    Args:
        slug: The document's filename stem, in kebab case.

    Returns:
        A title-cased name with minor words kept lowercase.
    """
    words = slug.replace("-", " ").split()
    return " ".join(
        word.capitalize() if index == 0 or word not in _TITLE_MINOR_WORDS else word
        for index, word in enumerate(words)
    )


def _title_from_body(body: list[str]) -> str:
    for raw in body:
        match = _ATX_HEADING_RE.match(raw)
        if match and len(match.group(1)) == 1:
            return match.group(2).strip()
    return ""


# Doctrine priority-table name -> framework filename stem.
#
# This table is explicit rather than fuzzy-matched on purpose. The doctrine
# table orders safety modes, with Crisis highest, so a near-miss string match
# that silently mis-ranked a framework would be a safety-relevant bug in a
# place no test would obviously catch. Every entry below was verified against
# the file each detector in `src/soulmap/runtime/detectors/` actually loads.
#
# Names absent from this map are absent on purpose:
#   crisis, dependency, mirror   safety modes and the default posture, with no
#                                framework document of their own
#   soulmate longing,            live under `skills/soulmate/`, which the
#   partnership patterns         allowlist does not publish
_TIER_ALIASES: dict[str, str] = {
    "de-escalation / sanctuary": "emotional-deescalation",
    "de-escalation": "emotional-deescalation",
    "grief": "grief-companion",
    "existential": "existential-companion",
    "direction": "life-direction",
    "shadow": "shadow-patterns",
    "sacred polarity": "sacred-feminine-masculine",
    "integration and celebration": "integration-celebration",
    "synthesis": "conversation-synthesis",
    "pattern": "pattern-mapper",
    "dark night of the soul": "dark-night-of-soul",
    "fear of visibility": "fear-of-visibility",
}


def _match_tier(slug: str, tiers: dict[str, str]) -> str | None:
    """Look up a framework document's doctrine priority tier.

    Args:
        slug: The document's filename stem.
        tiers: Mapping from ``load_priority_tiers``.

    Returns:
        The tier, or None when the document has no row in the doctrine table.
        None is a real answer, not a failure: several framework documents are
        topic lenses applied after a primary framework rather than routed
        primaries, and the build reports the rest as gaps.
    """
    for table_name, tier in tiers.items():
        if _TIER_ALIASES.get(table_name, table_name.replace(" ", "-")) == slug:
            return tier
    return None


def build_page(
    path: Path,
    repo_root: Path,
    tiers: dict[str, str],
    *,
    category: str,
) -> PublicPage:
    """Build one public page model from a canonical Markdown file.

    Internal sections are dropped here, before rendering, so they never enter
    the model a template can see.

    Args:
        path: The canonical source file.
        repo_root: Repository root, used for the source path.
        tiers: Priority tiers from ``load_priority_tiers``.
        category: The page's category, for example ``"frameworks"``.

    Returns:
        The page model.

    Raises:
        OSError: If the file cannot be read.
    """
    front, body = _split_front_matter(path.read_text(encoding="utf-8"))
    slug = path.stem
    name = front.get("name", "").strip() or _title_from_body(body) or slug
    if name == slug:
        # Front matter carries the kebab-case identifier rather than a display
        # name, so fall back to the document's own H1 before inventing one.
        name = _title_from_body(body) or _display_name(slug)

    renderer = _markdown_renderer()
    sections: list[PublicSection] = []
    for level, heading, lines in _drop_internal_blocks(_iter_blocks(body)):
        text = _delink_private_targets(
            "\n".join(_scrub_labelled_signals(lines)).strip()
        )
        if not text and not heading:
            continue
        sections.append(
            PublicSection(
                heading=heading,
                level=level or 2,
                anchor=slugify_github_anchor(heading) if heading else "",
                html=renderer.render(text) if text else "",
            )
        )

    return PublicPage(
        slug=slug,
        name=name,
        description=front.get("description", "").strip(),
        category=category,
        tier=_match_tier(slug, tiers) if category == "frameworks" else None,
        sections=tuple(sections),
        source_path=path.relative_to(repo_root).as_posix(),
    )


def load_public_pages(repo_root: Path) -> list[PublicPage]:
    """Load every allowlisted document as a public page model.

    Args:
        repo_root: Repository root.

    Returns:
        Page models in category then slug order. Documents that the allowlist
        does not name are never opened.

    Raises:
        OSError: If an allowlisted file cannot be read.
    """
    tiers = load_priority_tiers(repo_root)
    skills_root = repo_root / "skills"
    pages: list[PublicPage] = []

    for category_dir in sorted(skills_root.iterdir()):
        if not category_dir.is_dir():
            continue
        category = category_dir.name
        for path in sorted(category_dir.glob("*.md")):
            if not is_public_skill(category, path.name):
                continue
            pages.append(build_page(path, repo_root, tiers, category=category))

    return pages

File: site/src/content.ts
/**
 * Read SoulMap content for the public site.
 *
 * The site publishes the shipped knowledge package and nothing else. That set
 * is the root `SKILL.md`, `SOULMAP.md`, and everything under `skills/`, which
 * is exactly what the Python packager puts in the distribution archives.
 * `docs/`, `templates/`, `.claude/`, `tests/`, `scripts/`, and `src/` are
 * repository-only and must never reach a public page.
 *
 * The boundary is enforced here rather than trusted to the templates: every
 * document this module returns has already been checked against the allowed
 * roots, so a page cannot render something the packager would not ship.
 */

import { readFileSync, readdirSync, statSync } from "node:fs";
import { join, relative, sep } from "node:path";
import matter from "gray-matter";

/** Directories and files the site is allowed to publish. */
const PUBLISHABLE_ROOTS = ["skills"] as const;
const PUBLISHABLE_FILES = ["SKILL.md", "SOULMAP.md"] as const;

/** Repository-only paths. Present for an explicit check, not just omission. */
const INTERNAL_ROOTS = [
  "docs",
  "templates",
  ".claude",
  ".claude-plugin",
  ".github",
  "tests",
  "scripts",
  "src",
  "library",
  "evals",
  "dist",
] as const;

export interface SkillDocument {
  /** Repository-relative POSIX path, for example `skills/safety/SKILL.md`. */
  readonly path: string;
  /** URL slug, for example `safety/boundaries-safety`. */
  readonly slug: string;
  /** Front-matter `name`. */
  readonly name: string;
  /** Front-matter `description`. */
  readonly description: string;
  /** Markdown body with the front matter removed. */
  readonly body: string;
}

export interface Skill {
  /** Directory name, which the front matter `name` must match. */
  readonly id: string;
  readonly name: string;
  readonly description: string;
  readonly version: string;
  /** The skill's own `SKILL.md`. */
  readonly manifest: SkillDocument;
  /** Supporting documents inside the skill, excluding the manifest. */
  readonly documents: readonly SkillDocument[];
}

export interface SiteContent {
  readonly version: string;
  readonly doctrine: SkillDocument;
  readonly rootManifest: SkillDocument;
  readonly skills: readonly Skill[];
}

/**
 * Report whether a repository-relative path may be published.
 *
 * Exported because the boundary deserves a direct test rather than only being
 * exercised through a full build.
 */
export function isPublishable(relativePath: string): boolean {
  const normalised = relativePath.split(sep).join("/");
  if (normalised.startsWith("../") || normalised.includes("/../")) return false;

  const [head] = normalised.split("/");
  if (head === undefined) return false;
  if (INTERNAL_ROOTS.includes(head as (typeof INTERNAL_ROOTS)[number])) {
    return false;
  }
  if (PUBLISHABLE_FILES.includes(normalised as (typeof PUBLISHABLE_FILES)[number])) {
    return true;
  }
  return PUBLISHABLE_ROOTS.includes(head as (typeof PUBLISHABLE_ROOTS)[number]);
}

function readDocument(repoRoot: string, relativePath: string): SkillDocument {
  if (!isPublishable(relativePath)) {
    throw new Error(
      `Refusing to publish a repository-only path: ${relativePath}. ` +
        `The site may only render the shipped knowledge package.`,
    );
  }
  const raw = readFileSync(join(repoRoot, relativePath), "utf8");
  const { data, content } = matter(raw);
  const slug = relativePath.replace(/^skills\//, "").replace(/\.md$/, "");
  return {
    path: relativePath,
    slug,
    name: String(data["name"] ?? "").trim(),
    description: String(data["description"] ?? "").trim(),
    body: content.trim(),
  };
}

function markdownFilesIn(directory: string): string[] {
  return readdirSync(directory)
    .filter((entry) => entry.endsWith(".md"))
    .sort();
}

/** Read the package version from `pyproject.toml`, the single source of truth. */
function packageVersion(repoRoot: string): string {
  const text = readFileSync(join(repoRoot, "pyproject.toml"), "utf8");
  const match = /^version\s*=\s*"([^"]+)"/m.exec(text);
  if (!match?.[1]) {
    throw new Error("pyproject.toml does not declare project.version");
  }
  return match[1];
}

/** Load every publishable document, grouped by skill. */
export function loadContent(repoRoot: string): SiteContent {
  const skillsDir = join(repoRoot, "skills");
  const skills: Skill[] = readdirSync(skillsDir)
    .filter((entry) => statSync(join(skillsDir, entry)).isDirectory())
    .sort()
    .map((id) => {
      const manifest = readDocument(repoRoot, `skills/${id}/SKILL.md`);
      const documents = markdownFilesIn(join(skillsDir, id))
        .filter((file) => file !== "SKILL.md")
        .map((file) => readDocument(repoRoot, `skills/${id}/${file}`));

      const raw = readFileSync(join(skillsDir, id, "SKILL.md"), "utf8");
      const version = String(matter(raw).data["version"] ?? "").trim();

      return { id, name: manifest.name, description: manifest.description, version, manifest, documents };
    });

  return {
    version: packageVersion(repoRoot),
    doctrine: readDocument(repoRoot, "SOULMAP.md"),
    rootManifest: readDocument(repoRoot, "SKILL.md"),
    skills,
  };
}

/** Resolve the repository root from this file's location. */
export function repositoryRoot(fromUrl: string): string {
  const here = new URL(".", fromUrl).pathname;
  return relative(process.cwd(), join(here, "..", "..")) || ".";
}
